mgmt2/rpc/jsonrpc/JsonRpc.cc

```diff
@@ -0,0 +1,261 @@
+/**
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#include "JsonRpc.h"
+
+#include <iostream>
+#include <chrono>
+#include <system_error>
+
+#include "json/YAMLCodec.h"
+
+namespace rpc
+{
+static constexpr auto logTag = "rpc";
+
+namespace error = jsonrpc::error;
+
+JsonRpc::Dispatcher::response_type
+JsonRpc::Dispatcher::dispatch(jsonrpc::RpcRequestInfo const &request) const
+{
+  if (request.is_notification()) {
+    std::error_code ec;
+    invoke_notification_handler(request, ec);
+    if (ec) {
+      return {std::nullopt, ec};
+    }
+  } else {
+    return invoke_handler(request);
+  }
+
+  return response_type{};
+}
+
+std::pair<jsonrpc::MethodHandler, bool>
+JsonRpc::Dispatcher::find_handler(const std::string &methodName) const
+{
+  std::lock_guard<std::mutex> lock(_mutex);
+
+  const auto found = std::find_if(std::begin(_methods), std::end(_methods), [&](auto const &p) { return p.first == methodName; });
+  return found != std::end(_methods) ? std::make_pair(found->second, true) : std::make_pair(nullptr, false);
```

Comment:
`nullptr` doesn't suffice as a failure marker?

Reply:
it should.